File: py/wlearn/bundle.py
import hashlib
import json
import struct

from .errors import BundleError

BUNDLE_MAGIC = b'WLRN'
BUNDLE_VERSION = 1
HEADER_SIZE = 16


def _stable_json(obj):
    """Deterministic JSON: sorted keys, no whitespace."""
    return json.dumps(obj, sort_keys=True, separators=(',', ':')).encode('utf-8')


def encode_bundle(manifest, artifacts):
    """Encode a wlearn bundle.

    Args:
        manifest: dict with at least 'typeId'
        artifacts: list of {'id': str, 'data': bytes, 'mediaType'?: str}

    Returns:
        bytes
    """
    sorted_arts = sorted(artifacts, key=lambda a: a['id'])

    blob_offset = 0
    toc = []
    blobs = []

    for art in sorted_arts:
        data = art['data']
        if isinstance(data, memoryview):
            data = bytes(data)
        sha = hashlib.sha256(data).hexdigest()
        entry = {
            'id': art['id'],
            'offset': blob_offset,
            'length': len(data),
            'sha256': sha,
        }
        if 'mediaType' in art:
            entry['mediaType'] = art['mediaType']
        toc.append(entry)
        blobs.append(data)
        blob_offset += len(data)

    full_manifest = {**manifest, 'bundleVersion': BUNDLE_VERSION}
    manifest_bytes = _stable_json(full_manifest)
    toc_bytes = _stable_json(toc)

    header = struct.pack('<4sIII', BUNDLE_MAGIC, BUNDLE_VERSION,
                         len(manifest_bytes), len(toc_bytes))

    parts = [header, manifest_bytes, toc_bytes]
    parts.extend(blobs)
    return b''.join(parts)


def decode_bundle(data):
    """Decode a wlearn bundle.

    Args:
        data: bytes or memoryview

    Returns:
        (manifest: dict, toc: list, blobs: memoryview)
    """
    if isinstance(data, memoryview):
        buf = data
    else:
        buf = memoryview(data)

    if len(buf) < HEADER_SIZE:
        raise BundleError(
            f'Bundle too small: {len(buf)} bytes (minimum {HEADER_SIZE})')

    magic = bytes(buf[:4])
    if magic != BUNDLE_MAGIC:
        raise BundleError('Invalid bundle magic (expected WLRN)')

    version, manifest_len, toc_len = struct.unpack_from('<III', buf, 4)
    if version != BUNDLE_VERSION:
        raise BundleError(
            f'Unsupported bundle version: {version} (expected {BUNDLE_VERSION})')

    if HEADER_SIZE + manifest_len + toc_len > len(buf):
        raise BundleError(
            f'Bundle truncated: header declares {HEADER_SIZE + manifest_len + toc_len} '
            f'bytes but got {len(buf)}')

    try:
        manifest = json.loads(bytes(buf[HEADER_SIZE:HEADER_SIZE + manifest_len]))
    except json.JSONDecodeError as e:
        raise BundleError(f'Invalid manifest JSON: {e}') from e

    try:
        toc = json.loads(bytes(
            buf[HEADER_SIZE + manifest_len:HEADER_SIZE + manifest_len + toc_len]))
    except json.JSONDecodeError as e:
        raise BundleError(f'Invalid TOC JSON: {e}') from e

    blob_start = HEADER_SIZE + manifest_len + toc_len
    blob_region_len = len(buf) - blob_start

    # validate TOC entries
    for i, entry in enumerate(toc):
        offset = entry['offset']
        length = entry['length']
        if offset < 0 or length < 0 or offset + length > blob_region_len:
            raise BundleError(
                f'TOC entry "{entry["id"]}" out of bounds: '
                f'offset={offset}, length={length}, blobRegion={blob_region_len}')
        for j in range(i + 1, len(toc)):
            other = toc[j]
            a_start, a_end = offset, offset + length
            b_start, b_end = other['offset'], other['offset'] + other['length']
            if a_start < b_end and b_start < a_end and length > 0 and other['length'] > 0:
                raise BundleError(
                    f'TOC entries "{entry["id"]}" and "{other["id"]}" overlap')

    blobs = buf[blob_start:]
    return manifest, toc, blobs


def validate_bundle(data):
    """Decode and verify SHA-256 hashes of all blobs.

    Returns:
        (manifest, toc, blobs)
    """
    manifest, toc, blobs = decode_bundle(data)

    for entry in toc:
        blob = bytes(blobs[entry['offset']:entry['offset'] + entry['length']])
        actual = hashlib.sha256(blob).hexdigest()
        if actual != entry['sha256']:
            raise BundleError(
                f'SHA-256 mismatch for "{entry["id"]}": '
                f'expected {entry["sha256"]}, got {actual}')

    return manifest, toc, blobs

File: packages/core/src/bundle.js
import { BUNDLE_MAGIC, BUNDLE_VERSION, HEADER_SIZE } from '@wlearn/types'
import { BundleError } from './errors.js'
import { sha256Sync } from './hash.js'

// Deterministic JSON: sorted keys recursively, no whitespace, array order preserved
function stableStringify(val) {
  if (val === null || val === undefined) return JSON.stringify(val)
  if (typeof val !== 'object') return JSON.stringify(val)
  if (Array.isArray(val)) {
    return '[' + val.map(v => stableStringify(v)).join(',') + ']'
  }
  const keys = Object.keys(val).sort()
  return '{' + keys.map(k => JSON.stringify(k) + ':' + stableStringify(val[k])).join(',') + '}'
}

const textEncoder = new TextEncoder()
const textDecoder = new TextDecoder()

/**
 * Encode a wlearn bundle (WLRN format).
 *
 * @param {Object} manifest - Bundle manifest. Must include `typeId` (e.g. `'wlearn.xgboost.classifier@1'`).
 *   May include `params`, `requires`, `seed`, or any model-specific metadata.
 * @param {Array<{id: string, mediaType?: string, data: Uint8Array}>} artifacts - Artifact blobs.
 *   Each entry has `id` (unique within the bundle), optional `mediaType`, and `data` (raw bytes).
 *   Artifacts are sorted by `id` for determinism; SHA-256 hashes are computed automatically.
 * @returns {Uint8Array} The encoded bundle bytes (header + manifest JSON + TOC JSON + blob region).
 */
export function encodeBundle(manifest, artifacts) {
  // Sort artifacts by id for determinism
  const sorted = [...artifacts].sort((a, b) => a.id < b.id ? -1 : a.id > b.id ? 1 : 0)

  // Build TOC and compute blob region
  let blobOffset = 0
  const toc = []
  const blobs = []

  for (const art of sorted) {
    const data = art.data instanceof Uint8Array ? art.data : new Uint8Array(art.data)
    const hash = sha256Sync(data)
    const entry = { id: art.id, offset: blobOffset, length: data.length, sha256: hash }
    if (art.mediaType) entry.mediaType = art.mediaType
    toc.push(entry)
    blobs.push(data)
    blobOffset += data.length
  }

  const fullManifest = { ...manifest, bundleVersion: BUNDLE_VERSION }
  const manifestBytes = textEncoder.encode(stableStringify(fullManifest))
  const tocBytes = textEncoder.encode(stableStringify(toc))

  const totalLen = HEADER_SIZE + manifestBytes.length + tocBytes.length + blobOffset
  const out = new Uint8Array(totalLen)
  const view = new DataView(out.buffer)

  // Header
  out.set(BUNDLE_MAGIC, 0)
  view.setUint32(4, BUNDLE_VERSION, true)
  view.setUint32(8, manifestBytes.length, true)
  view.setUint32(12, tocBytes.length, true)

  // Manifest + TOC + blobs
  out.set(manifestBytes, HEADER_SIZE)
  out.set(tocBytes, HEADER_SIZE + manifestBytes.length)

  let pos = HEADER_SIZE + manifestBytes.length + tocBytes.length
  for (const blob of blobs) {
    out.set(blob, pos)
    pos += blob.length
  }

  return out
}

/**
 * Decode a wlearn bundle (WLRN format).
 *
 * @param {Uint8Array|ArrayBuffer} bytes - Raw bundle bytes.
 * @returns {{manifest: Object, toc: Array<{id: string, offset: number, length: number, sha256: string, mediaType?: string}>, blobs: Uint8Array}}
 *   `manifest` is the parsed manifest object. `toc` is an array of blob descriptors.
 *   `blobs` is the concatenated blob region -- slice individual blobs with
 *   `blobs.subarray(entry.offset, entry.offset + entry.length)`.
 * @throws {BundleError} On invalid magic, unsupported version, truncated data, or malformed JSON.
 */
export function decodeBundle(bytes) {
  const buf = bytes instanceof Uint8Array ? bytes : new Uint8Array(bytes)

  if (buf.length < HEADER_SIZE) {
    throw new BundleError(`Bundle too small: ${buf.length} bytes (minimum ${HEADER_SIZE})`)
  }

  // Verify magic
  for (let i = 0; i < 4; i++) {
    if (buf[i] !== BUNDLE_MAGIC[i]) {
      throw new BundleError('Invalid bundle magic (expected WLRN)')
    }
  }

  const view = new DataView(buf.buffer, buf.byteOffset, buf.byteLength)
  const version = view.getUint32(4, true)
  if (version !== BUNDLE_VERSION) {
    throw new BundleError(`Unsupported bundle version: ${version} (expected ${BUNDLE_VERSION})`)
  }

  const manifestLen = view.getUint32(8, true)
  const tocLen = view.getUint32(12, true)

  if (HEADER_SIZE + manifestLen + tocLen > buf.length) {
    throw new BundleError(`Bundle truncated: header declares ${HEADER_SIZE + manifestLen + tocLen} bytes but got ${buf.length}`)
  }

  let manifest, toc
  try {
    manifest = JSON.parse(textDecoder.decode(buf.subarray(HEADER_SIZE, HEADER_SIZE + manifestLen)))
  } catch (e) {
    throw new BundleError(`Invalid manifest JSON: ${e.message}`)
  }

  try {
    toc = JSON.parse(textDecoder.decode(buf.subarray(HEADER_SIZE + manifestLen, HEADER_SIZE + manifestLen + tocLen)))
  } catch (e) {
    throw new BundleError(`Invalid TOC JSON: ${e.message}`)
  }

  const blobStart = HEADER_SIZE + manifestLen + tocLen
  const blobRegionLen = buf.length - blobStart

  // Validate TOC entries: no overlaps, within bounds
  for (let i = 0; i < toc.length; i++) {
    const entry = toc[i]
    if (entry.offset < 0 || entry.length < 0 || entry.offset + entry.length > blobRegionLen) {
      throw new BundleError(`TOC entry "${entry.id}" out of bounds: offset=${entry.offset}, length=${entry.length}, blobRegion=${blobRegionLen}`)
    }
    for (let j = i + 1; j < toc.length; j++) {
      const other = toc[j]
      const aStart = entry.offset, aEnd = entry.offset + entry.length
      const bStart = other.offset, bEnd = other.offset + other.length
      if (aStart < bEnd && bStart < aEnd && entry.length > 0 && other.length > 0) {
        throw new BundleError(`TOC entries "${entry.id}" and "${other.id}" overlap`)
      }
    }
  }

  const blobs = buf.subarray(blobStart)
  return { manifest, toc, blobs }
}

/**
 * Encode a value as deterministic JSON bytes (sorted keys, no whitespace).
 * @param {*} obj - Value to serialize.
 * @returns {Uint8Array} UTF-8 encoded JSON bytes.
 */
export function encodeJSON(obj) {
  return textEncoder.encode(stableStringify(obj))
}

/**
 * Decode UTF-8 JSON bytes to a JS value.
 * @param {Uint8Array|ArrayBuffer} bytes - UTF-8 encoded JSON.
 * @returns {*} Parsed value.
 */
export function decodeJSON(bytes) {
  const buf = bytes instanceof Uint8Array ? bytes : new Uint8Array(bytes)
  return JSON.parse(textDecoder.decode(buf))
}

/**
 * Decode and validate a bundle, verifying SHA-256 hashes of all blobs.
 * @param {Uint8Array|ArrayBuffer} bytes - Raw bundle bytes.
 * @returns {{manifest: Object, toc: Array, blobs: Uint8Array}} Same shape as `decodeBundle`.
 * @throws {BundleError} On format errors or hash mismatch.
 */
export function validateBundle(bytes) {
  const { manifest, toc, blobs } = decodeBundle(bytes)

  for (const entry of toc) {
    const blob = blobs.subarray(entry.offset, entry.offset + entry.length)
    const hash = sha256Sync(blob)
    if (hash !== entry.sha256) {
      throw new BundleError(`SHA-256 mismatch for "${entry.id}": expected ${entry.sha256}, got ${hash}`)
    }
  }

  return { manifest, toc, blobs }
}
